### PR TITLE
Upgrade to Go 1.17 and switch to Go modules and Github actions

### PR DESCRIPTION
**Pull request by speijnik**
This PR upgrades Go to 1.17, switches to Go modules for dependency management and Github actions for CI

### CODE REVIEW

File: tools/go.mod [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module go.anx.io/go-anxcloud/tools
```
Comment (LittleFox94) [inline tools/go.mod]: the module has to be changed

File: Makefile [hunk 2]
```diff
@@ -1,52 +1,95 @@
...
+test: tools
+	tools/ginkgo run -p 			\
+		-timeout 0 					\
+		-race 						\
+		-coverprofile coverage.out 	\
+		./...
+	go tool cover -html=coverage.out -o coverage.html
+
+.PHONY: func-test
+func-test: tools
```
Comment (LittleFox94) [inline Makefile]: I think you can strip this entire target. It's used in go-anxcloud for integration tests against the Engine API, nothing this library will ever have.

File: README.md [hunk 3]
```diff
@@ -3,17 +3,17 @@ go-structmapper
...
 [![GoDoc](https://godoc.org/gopkg.in/anexia-it/go-structmapper.v1?status.svg)](https://godoc.org/gopkg.in/anexia-it/go-structmapper.v1)
-[![Build Status](https://travis-ci.org/anexia-it/go-structmapper.svg?branch=v1)](https://travis-ci.org/anexia-it/go-structmapper)
 [![codecov](https://codecov.io/gh/anexia-it/go-structmapper/branch/v1/graph/badge.svg)](https://codecov.io/gh/anexia-it/go-structmapper)
 [![Go Report Card](https://goreportcard.com/badge/gopkg.in/anexia-it/go-structmapper.v1)](https://goreportcard.com/report/gopkg.in/anexia-it/go-structmapper.v1)
 
 go-structmapper is a Go library which allows mapping arbitrary structs to maps
 and vice-versa.
 
-## Install
+Install
```
Comment (LittleFox94) [inline README.md]: Why the different syntax? The previous one is the preferred markdown syntax

File: common_test.go [hunk 4]
```diff
@@ -5,7 +5,7 @@ import (
 
 	"net"
 
-	"github.com/anexia-it/go-structmapper"
+	structmapper "github.com/anexia-it/go-structmapper"
```
Comment (LittleFox94) [inline common_test.go]: you don't need this change

File: Makefile [hunk 5]
```diff
@@ -1,52 +1,95 @@
...
+.PHONY: benchmark
 benchmark:
-	@echo "Running benchmarks..."
-	@$(foreach pkg,$(PACKAGES), \
-		go test -bench="." $(GO_BENCHMARK_FLAGS) $(pkg);)
+	go test -bench=. -benchmem ./...
+
+.PHONY: test
+test: tools
+	tools/ginkgo run -p 			\
```
Comment (LittleFox94) [inline Makefile]: using the ginkgo runner while not using ginkgo smells like a unneeded dependency, using `go test -race ./...` probably works the same for this project

File: README.md [hunk 6]
```diff
@@ -3,17 +3,17 @@ go-structmapper
 
 [![license](https://img.shields.io/github/license/mashape/apistatus.svg?maxAge=2592000)](https://github.com/anexia-it/go-structmapper/blob/master/LICENSE)
 [![GoDoc](https://godoc.org/gopkg.in/anexia-it/go-structmapper.v1?status.svg)](https://godoc.org/gopkg.in/anexia-it/go-structmapper.v1)
-[![Build Status](https://travis-ci.org/anexia-it/go-structmapper.svg?branch=v1)](https://travis-ci.org/anexia-it/go-structmapper)
```
Comment (LittleFox94) [inline README.md:6]: GitHub actions also give you a status badge, maybe change to that instead of removing it?

**Review by LittleFox94 — CHANGES_REQUESTED** [review]
You can also remove the `Gopkg.yaml`/`Gopkg.lock` now.

On a general note, using a `main` branch instead of `v1` would probably be good, Go does not resolve the major version via the branch name.

**Comment by speijnik** [comment]
@LittleFox94 thanks for the initial review. I believe I addressed all of your comments.
As for the default branch: the idea is to branch main off of v1 once these changes have been merged, so this is planned already.

File: tools/.gitignore [hunk 7]
```diff
@@ -0,0 +1,4 @@
+/tools
+/ginkgo
```
Comment (LittleFox94) [inline tools/.gitignore:2]: can be removed

Comment (LittleFox94) [inline tools/.gitignore]: can be removed

File: tools/.gitignore [hunk 8]
```diff
@@ -0,0 +1,4 @@
+/tools
```
Comment (LittleFox94) [inline tools/.gitignore]: can be removed, go-anxcloud specific

File: Makefile [hunk 9]
```diff
@@ -1,52 +1,82 @@
...
+	go tool cover -html=coverage.out -o coverage.html
+
+.PHONY: go-lint
+go-lint: tools
+	@echo "==> Checking source code against linters..."
+	@tools/golangci-lint run ./...
+	@tools/golangci-lint run --build-tags integration ./...
+
+.PHONY: docs-lint
+docs-lint: tools
```
Comment (LittleFox94) [inline Makefile:42]: only now realized you didn't copy the `documentation` workflow over, which uses this target. I'd advise to copy that workflow, too.

**Review by LittleFox94 — APPROVED** [review]
Looks good, found some minor things you can decide yourself if worth fixing or not :)

A note to the GitHub Actions workflows: I want to refactor them in go-anxcloud to be (mostly) a single workflow with multiple jobs, maybe you want to merge them into a single workflow, too. Due to the integration tests it isn't done in go-anxcloud, yet.

File: Makefile [hunk 10]
```diff
@@ -1,6 +1,6 @@
 GITTAG ?= $(shell git describe --tags --always)
 GITCOMMIT ?= $(shell git log -1 --pretty=format:"%H")
-GOLDFLAGS ?= -s -w -extldflags '-zrelro -znow' -X go.anx.io/go-anxcloud/pkg/client.version=$(GITTAG)
```
Comment (LittleFox94) [inline Makefile]: good catch!

**Comment by speijnik** [comment]
@LittleFox94 I would re-request a review, but technically you already did one for my latest changes.
Thanks for the reviews!

**Comment by LittleFox94** [comment]
Gladly :)

**Pull request merged**